File: Novomber 2022/Day 16/C++ Code/Sum of Beauty of All Substrings.cpp
// Problem Link :- https://practice.geeksforgeeks.org/problems/sum-of-beauty-of-all-substrings-1662962118/1
// Video Solution Link :- 

//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

class Solution {
  public:
    // Time Complexity :- O(n*n*26) => O(n^2);
    // space Complexity :- O(26) => O(1);
    
    int beautySum(string s) {
        // Your code goes here
        int ans=0;
        for(int i=0;i<s.length();i++) {
            vector<int> fre(26,0);
            for(int j=i;j<s.length();j++) {
                fre[s[j]-'a']++;
                ans+=find(fre);
            }
        }
        return ans;
    }
    
    int find(vector<int> fre) {
        int mx=INT_MIN,mn=INT_MAX;
        for(int f : fre) {
            if(f==0) continue;
            mx = max(mx,f);
            mn = min(mn,f);
        }
        return mx-mn;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        string s;
        cin >> s;
        Solution obj;
        cout << obj.beautySum(s) << endl;
    }
    return 0;
}
// } Driver Code Ends

File: Novomber 2022/Day 16/JAVA Code/Sum of Beauty of All Substrings.java
// Problem Link :- https://practice.geeksforgeeks.org/problems/sum-of-beauty-of-all-substrings-1662962118/1
// Video Solution Link :- 

//{ Driver Code Starts
import java.io.*;
import java.util.*;

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t;
        t = Integer.parseInt(br.readLine());
        while(t-- > 0){
            //br.readLine();
            String s;
            s = br.readLine();
            
            Solution obj = new Solution();
            int res = obj.beautySum(s);
            
            System.out.println(res);
            
        }
    }
}

// } Driver Code Ends


class Solution {
    public static int beautySum(String s) {
        // code here
        int ans=0;
        for(int i=0;i<s.length();i++) {
            int[] fre = new int[26];
            for(int j=i;j<s.length();j++) {
                fre[s.charAt(j)-'a']++;
                ans+=findBeauty(fre);
            }
        }
        return ans;
    }
    
    public static int findBeauty(int[] fre) {
        int max=Integer.MIN_VALUE,min=Integer.MAX_VALUE;
        for(int f : fre) {
            if(f==0) continue;
            max = Math.max(max,f);
            min = Math.min(min,f);
        }
        return max-min;
    }
}
        
